---
title: "Loss of Arabidopsis Matrix metalloproteinase-5 affects root development and root bacterial communities during drought stress"
author: "Daniel Caddell"
date: "8/21/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## R Markdown

## load packages
```{r}
library(phyloseq); packageVersion("phyloseq") #‘1.30.0’
library(ggplot2); packageVersion("ggplot2") #‘3.2.1’
library(reshape2); packageVersion("reshape2") #‘1.4.3’
library(dplyr); packageVersion("dplyr") #‘0.8.3’
library(RColorBrewer); packageVersion("RColorBrewer") #‘1.1.2’
library(beepr); packageVersion("beepr") #‘1.3’
library(scales); packageVersion("scales") #‘1.1.0’
library(grid); packageVersion("grid") #‘3.6.1’
library(ape); packageVersion("ape") #‘5.3’
library(ggalt); packageVersion("ggalt") #‘0.4.0’
library(mctoolsr); packageVersion("mctoolsr") # ‘0.1.1.2’
library(vegan); packageVersion("vegan") # ‘2.5.6’
library(data.table); packageVersion("data.table") #‘1.12.8’
library(agricolae);packageVersion("agricolae") #‘1.3.1’
library(patchwork);packageVersion("patchwork") #‘1.0.0’
theme_set(theme_bw()) #Define a default theme for ggplot graphics.

#setwd("/set/working/directory/location/")
setwd("/users/dcaddell/Desktop/Laxmi/mmp5/")
```

## import OTU biom file and sample file, then build phyloseq object  
```{r}
biom_file <- "merged-feature-table.biom"
map_file <- "metadata_MMP5.txt"
otutax <- import_biom(biom_file, parseFunction = parse_taxonomy_default)
sam <- import_qiime_sample_data(map_file)
all <- merge_phyloseq(otutax, sam)
all
colnames(tax_table(all)) <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")
colnames(tax_table(all))
```

## Remove mitochondrial and chloroplast reads from a phyloseq object (Illumina - greengenes)  
A full list of related functions are in 'functions_for16SData.Rmd'  
WARNING: only use on taxonomies generated by greengenes, otherwise pattern matching will not find mito/chloro ASVs/OTUs
```{r}
remove_mitoAndChloro <- function(x){#where x is a merged phyloseq object
  df <- as.data.frame(tax_table(x))
  setDT(df, keep.rownames = TRUE)
  colnames(df)[1] <- "OTU"
  allASVs <- df[,1]
  mito <- filter(df, df$Family == "f__mitochondria")
  chloro <- filter(df, df$Class == "c__Chloroplast")
  contam <- as.vector(rbind(mito, chloro)$OTU)
  ASVs_toKeep <- as.vector(subset(allASVs, !(OTU %in% contam))$OTU)
  num_mito <- nrow(mito)
  num_chloro <- nrow(chloro)
  cat(sprintf("%i and %s ASVs were removed as pesky mitochondria or chloroplast, respectively.\nKhaleesi frees these bacterial slaves to the wild binary yonder.", num_mito, num_chloro))
  x <- prune_taxa(ASVs_toKeep, x)
}
```

## Clean up taxonomy
Takes a psmelted phyloseq object and cleans up greengenes-format taxonomies.  
i.e. removes the "k__" in k__Bacteria  
Also takes NA values and unassigned taxa and renames their taxonomy as "Unknown"  
```{r}
cleanup_taxonomy <- function(x){ #to be used on dfs made from melted phyloseq objects
  x$Domain <- as.character(lapply(x$Domain, sub, pattern = "k__", replacement = ""))
  x$Phylum <- as.character(lapply(x$Phylum, sub, pattern = "p__", replacement = ""))
  x$Class <- as.character(lapply(x$Class, sub, pattern = "c__", replacement = ""))
  x$Order <- as.character(lapply(x$Order, sub, pattern = "o__", replacement = ""))
  x$Family <- as.character(lapply(x$Family, sub, pattern = "f__", replacement = ""))
  x$Genus <- as.character(lapply(x$Genus, sub, pattern = "g__", replacement = ""))
  x$Species <- as.character(lapply(x$Species, sub, pattern = "s__", replacement = ""))
  x$Domain <- ifelse(x$Domain == "", "Unknown", as.character(x$Domain))
  x$Phylum <- ifelse(x$Phylum == "", "Unknown", as.character(x$Phylum))
  x$Class <- ifelse(x$Class == "", "Unknown", as.character(x$Class))
  x$Order <- ifelse(x$Order == "", "Unknown", as.character(x$Order))
  x$Family <- ifelse(x$Family == "", "Unknown", as.character(x$Family))
  x$Genus <- ifelse(x$Genus == "", "Unknown", as.character(x$Genus))
  x$Species <- ifelse(x$Species == "", "Unknown", as.character(x$Species))
  x$Domain <- ifelse(is.na(x$Domain) == TRUE , "Unknown", as.character(x$Domain))
  x$Phylum <- ifelse(is.na(x$Phylum) == TRUE , "Unknown", as.character(x$Phylum))
  x$Class <- ifelse(is.na(x$Class) == TRUE , "Unknown", as.character(x$Class))
  x$Order <- ifelse(is.na(x$Order) == TRUE , "Unknown", as.character(x$Order))
  x$Family <- ifelse(is.na(x$Family) == TRUE , "Unknown", as.character(x$Family))
  x$Genus <- ifelse(is.na(x$Genus) == TRUE , "Unknown", as.character(x$Genus))
  x$Species <- ifelse(is.na(x$Species) == TRUE , "Unknown", as.character(x$Species))
  x <- x
}
```

Now run the cleanup functions to remove any remaining mitochondria and chloroplast sequences.  
Also melt phyloseq object into a dataframe and transform sample counts for further analyses.  
```{r}
all.clean <- remove_mitoAndChloro(all)
all.clean.df <- psmelt (all.clean)
all.clean.df <- cleanup_taxonomy(all.clean.df)
#transform data
all.trans <- transform_sample_counts(all.clean, function(OTU) 100*OTU/sum(OTU))
#merge groups within transformed data
all.trans.merge <- merge_samples(all.trans, "Group")
all.trans.merge <- transform_sample_counts(all.trans.merge, function(OTU) 100*OTU/sum(OTU))
all.tm.df<- psmelt(all.trans.merge)
all.tm.df <- cleanup_taxonomy(all.tm.df) #NOTE: some columns were messed up during merge_samples
beep(4)
```

## Save the session so that it can be loaded later without having to rerun this step.
```{r eval=FALSE, include=TRUE}
save.image("phyloseq_illumina.rdata")
#close R, re-open R
load("phyloseq_illumina.rdata")
```

## Alpha Diversity   
Note: this was performed with rarefied data, which differs from the transformed data above.
```{r}
set.seed(42)
all_rare<-rarefy_even_depth(all.clean, rngseed = TRUE)
#sanity check
par(mfrow = c(1, 2))
title = "Sum of reads for each sample, all_rare"
plot(sort(sample_sums(all_rare), TRUE), type = "h", main = title, ylab = "reads") 
#, ylim = c(0, 1500))

richness = estimate_richness(all_rare,measures = "Shannon")
richness
```

ANOVA of alpha-diversity (and Tukey-HSD posthoc test)
```{r}
temp.aov <- sample_data(all_rare)
data.aov <- cbind(temp.aov, richness)
data.aov 
#ANOVA posthoc test: Tukey-HSD
all_aov <- aov(Shannon ~ Group, data = data.aov)
summary(all_aov) #F-value = 0.177, alpha diversity is not significant!, no need for posthoc test
#all_aov_tukey <- TukeyHSD(all_aov, 'Group', conf.level=0.95)
#all_aov_tukey 

#Plot Shannon's diversity
p1d <- plot_richness(all_rare, x="Group", measures=c("Shannon"), color="Group")+
        geom_point( size = 6) +
        geom_point(shape = 1,size = 6,colour = "black")+
        ylim(4.3,6)+
        scale_color_manual(values=c("#444546","#4d5b6a","#a04344","#bf7a8f","#6a90b4","#a1c5c8"))+
        theme(axis.text.x=element_text(size=12,color="black",angle=90), 
              axis.text.y=element_text(size=12,color="black"), 
              axis.title=element_text(size=12,face="bold"), 
              text=element_text(size=12),
              legend.title=element_text(size=12), 
              legend.text=element_text(size=12))
p1d
#ggsave("alphaD.pdf", plot = p1d, width=6, height=3, useDingbats=FALSE)
#ggsave("alphaD.png", plot = p1d, width=6, height=3, dpi=600)
```

## Beta Diversity
CAP analysis  
```{r}
ordcap = ordinate(all.trans, "CAP", "bray", ~Treatment*Genotype)

  p2d <- phyloseq::plot_ordination(all.trans, ordcap, "samples", color="Group", title="Bray-Curtis distance")+
          geom_vline(xintercept = 0,linetype="longdash",color="gray")+
          geom_hline(yintercept = 0,linetype="longdash",color="gray")+
    geom_point( size = 7)+
          geom_point(shape = 1,size = 7,colour = "black")+
          scale_color_manual(values=c("#444546","#4d5b6a","#a04344","#bf7a8f","#6a90b4","#a1c5c8"))+
          scale_x_reverse()+
          theme(axis.text.x=element_text(size=11,color="black",angle=0), 
                axis.text.y=element_text(size=11,color="black"), 
                axis.title=element_text(size=11,face="bold"), 
                text=element_text(size=11, face="bold"),
                plot.title=element_text(size=15,hjust= 0.5, vjust=2),
                legend.title=element_text(size=10), 
                legend.text=element_text(size=8))
  p2d
#ggsave("betaD.pdf", plot = p2d, width=6, height=5, useDingbats=FALSE)
#ggsave("betaD.png", plot = p2d, width=6, height=5, dpi=600)
```

adonis test and pairwise permanova posthoc test
```{r}
ps.prop_dist <- distance(all.trans, method = "bray")
ps.prop_data <- data.frame(sample_data(all.trans))
summary(ps.prop_data)

test <- adonis(ps.prop_dist ~ Group, data = ps.prop_data, permutations = 1000)
test
# get test results as a table
test_tab <- test$aov.tab
test_tab_rows <- rownames(test_tab)
test_tab_rows
for(i in 1:3){ #change "1:n" to number of rows being assessed
  a1 <- c("The percent of variance attributable to the factor")
  b2 <- toString(test_tab_rows[i])
  c3 <- c("is")
  d4 <- toString(round(test_tab$SumsOfSqs[i]/test_tab$SumsOfSqs[3], 3)) #change [n] to the row of "Total"
  print(paste(a1,b2,c3,d4))
}
#adonis posthoc test: Pairwise permanova
calc_pairwise_permanovas(ps.prop_dist, ps.prop_data, "Treatment")
calc_pairwise_permanovas(ps.prop_dist, ps.prop_data, "Genotype")
pm1 <- calc_pairwise_permanovas(ps.prop_dist, ps.prop_data, "Group")
pm1
#write.csv(pm1,"pm1.csv")

```


## Relative abundance
```{r}
# Agglomerate Taxa 
glom_Phylum <- tax_glom(all.trans, taxrank = "Phylum")
glom_Phylum
## pick a taxonomic level
taxrank_level = "Phylum"
gloms <- get(paste0("glom_", taxrank_level))
dim(otu_table(gloms))[1]
gloms
## get top 10 class
others <- names(sort(taxa_sums(gloms), decreasing = TRUE)[6:length(names(taxa_sums(gloms)))])
gloms <- merge_taxa(gloms, others, 1)
gloms
# sort orders based on relative abundance
# get otu table
ratab <- data.frame(otu_table(gloms))
dim(ratab)
# calculate average relative abundance for all the samples
ratab <- data.frame(rowMeans(ratab))
colnames(ratab) <- "Relative_Abundance"
# merge otu table with taxonomy table
ratab <- merge(as.data.frame(gloms@tax_table@.Data), ratab, by = "row.names")
# sorting taxa by abundance from high to low for RA
sort0 <- c()
sort0 <- ratab[with(ratab, order(ratab$Relative_Abundance)),][[taxrank_level]]
sort0 <- as.vector(sort0)
sort0[is.na(sort0)] <- "Other"
sort0 <- sort0[! sort0 %in% "Other"]
sort0 <- c("Other", sort0)
sort0
## melt dataframe
melts <- psmelt(gloms)
# convert Class to a character vector from a factor 
str(melts[[taxrank_level]])
melts[[taxrank_level]] <- as.character(melts[[taxrank_level]])
# replace all NAs with Other
melts[[taxrank_level]][is.na(melts[[taxrank_level]])] <- "Other"
# convert Class back to a factor
melts[[taxrank_level]] <- as.factor(melts[[taxrank_level]])
levels(factor(melts[[taxrank_level]]))
melts$Phylum <- factor(melts$Phylum, levels=sort0)
# sort taxa
summary(melts)
```

Plot relative abundance of the "top 5" phyla
```{r}
farrowAndBall_swatch <- c(
  "#4d5b6a" #Stiffkey Blue
  ,"#a1c5c8" #Blue Ground
  ,"#686a47" #Bancha
  ,"#ecc363" #Babouche
  ,"#a04344" #Incarnadine  
  ,"#50414c" #Pelt
)

# plot relative abundance
ra1 <- ggplot(data=melts, aes(x=Group, y=Abundance, fill=factor(get(taxrank_level)))) +
  geom_bar(stat="identity",position = "stack") +
  scale_fill_manual(values = farrowAndBall_swatch) +
  theme(axis.text.x=element_text(size=12,color="black",angle=90), 
        axis.text.y=element_text(size=12,color="black"), 
        axis.title=element_text(size=12,face="bold"),
        text=element_text(size=12)) +
  guides(fill=guide_legend(title=taxrank_level))
ra1
#ggsave("rel_abun.pdf", plot = ra1, width=6, height=8, useDingbats=FALSE)
#ggsave("rel_abun.png", plot = ra1, width=6, height=8, dpi=600)
```

Test significance of Actinobacteria between samples and plot.
```{r}
all.actino <- subset_taxa(all.trans, Phylum %in% c("p__Actinobacteria"))
all.Actino.ASV.temp <- psmelt(all.actino)
all.Actino.ASV <- subset(all.Actino.ASV.temp, Abundance > 0)
length(unique(all.Actino.ASV$OTU))

Actino.summary <- all.Actino.ASV%>%
  group_by(Group)%>%
  summarise(total=sum(Abundance))
Actino.summary
#write.csv(Actino.summary,"Actino.summary.csv")

#### REORGANIZED SAMPLES BACK INTO GROUPS PERFORMED OUTSIDE OF R
Actino.summary_RA <- read.csv("Actino.summary_RA.csv")
actino_aov <- aov(RA ~ Sample, data = Actino.summary_RA)
summary(actino_aov)
Actino_tukey <- HSD.test(actino_aov, 'Sample')
Actino_tukey

### Plotting Actinobacteria relative abundance during drought
p<- ggplot(data=Actino.summary_RA, aes(x=Group, y=RA_log2, fill=Group))+
    geom_boxplot(fill=c("#444546","#4d5b6a","#a04344","#bf7a8f","#6a90b4","#a1c5c8"))+
    geom_dotplot(binaxis='y',fill="#333333", stackdir='center', stackratio=1, dotsize=1.25)+
    labs(title = "Actinobacteria", x = "Group", y = "log2 abundance") +
    theme_bw(base_size = 20) +
    theme(axis.text.x=element_text(size=12,color="black",angle=90), 
        axis.text.y=element_text(size=12,color="black"), 
        axis.title=element_text(size=12,face="bold"),
        text=element_text(size=12)) +
    theme(plot.title = element_text(size = 20, hjust = 0.5))
p
#ggsave("Actino_RA.pdf", plot = p, width=6, height=7, useDingbats=FALSE)
#ggsave("Actino_RA.png", plot = p, width=6, height=7, dpi=600)
```
